### PR TITLE
fix(amazonq): chat @workspace file indexing respect git-ignore  

### PR DESCRIPTION
**Pull request by zixlin7**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Description
1. refactor file selection for workspace context to ignore git-ignore and other should ignored files
2. use new language server version

## Checklist
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/ProjectContextController.kt [hunk 1]
```diff
@@ -8,17 +8,16 @@ import com.intellij.openapi.components.Service
...
+import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.launch
 import software.aws.toolkits.core.utils.getLogger
 import software.aws.toolkits.core.utils.warn
-import software.aws.toolkits.jetbrains.core.coroutines.disposableCoroutineScope
 import software.aws.toolkits.jetbrains.services.codewhisperer.settings.CodeWhispererSettings
 
 @Service(Service.Level.PROJECT)
-class ProjectContextController(val project: Project) : Disposable {
+class ProjectContextController(val project: Project, val scope: CoroutineScope) : Disposable {
```
Comment (Will-ShaoHua) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/ProjectContextController.kt:18]: could be `private`

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/ProjectContextProvider.kt [hunk 2]
```diff
@@ -264,59 +262,35 @@ class ProjectContextProvider(val project: Project, private val encoderServer: En
...
-                    val current = stack.pop()
+        val changeListManager = ChangeListManager.getInstance(project)
+        val featureDevSessionContext = FeatureDevSessionContext(project)
+        val allFiles = project.guessProjectDir()?.let {
+            VfsUtil.collectChildrenRecursively(it).filter { child ->
+                val language = child.programmingLanguage()
+                !changeListManager.isIgnoredFile(child) && !isBuildOrBin(child.path) &&
+                    runBlocking { !featureDevSessionContext.ignoreFile(child, scope) } && child.length <= 10 * 1024 * 1024 &&
+                    language != CodeWhispererPlainText.INSTANCE && language != CodeWhispererUnknownLanguage.INSTANCE
+            }
```
Comment (Will-ShaoHua) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/ProjectContextProvider.kt:280]: nit
```suggestion
                !changeListManager.isIgnoredFile(child) && 
                !isBuildOrBin(child.path) &&
                 runBlocking { !featureDevSessionContext.ignoreFile(child, scope) } && 
                 child.length <= 10 * 1024 * 1024 &&
                 language !is CodeWhispererPlainText && 
                 language !is CodeWhispererUnknownLanguage
            }
```

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/ProjectContextProvider.kt [hunk 3]
```diff
@@ -264,59 +262,35 @@ class ProjectContextProvider(val project: Project, private val encoderServer: En
...
             }
+            collectedFiles.add(file.path)
+            currentTotalFileSize += file.length
         }
+
         return FileCollectionResult(
-            files = files.toList(),
-            fileSize = (currentTotalFileSize / 1024 / 102).toInt()
+            files = collectedFiles.toList(),
+            fileSize = (currentTotalFileSize / 1024 / 1024).toInt()
```
Comment (Will-ShaoHua) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/ProjectContextProvider.kt:293]: qq why dont we simply use byte

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/ProjectContextProvider.kt [hunk 4]
```diff
@@ -264,59 +262,35 @@ class ProjectContextProvider(val project: Project, private val encoderServer: En
...
+        val collectedFiles = mutableListOf<String>()
         var currentTotalFileSize = 0L
-        val stack = Stack<VirtualFile>()
-        moduleLoop@ for (module in project.modules) {
-            if (module.guessModuleDir() != null) {
-                stack.push(module.guessModuleDir())
-                while (stack.isNotEmpty()) {
-                    val current = stack.pop()
+        val changeListManager = ChangeListManager.getInstance(project)
+        val featureDevSessionContext = FeatureDevSessionContext(project)
```
Comment (leigaol) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/ProjectContextProvider.kt:273]: Might be better to not depend on FeatureDev but depend on a util function

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/ProjectContextProvider.kt (hunk 2 above)
Comment (leigaol) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/ProjectContextProvider.kt:280]: the index actually supports multiple languages include Config ( brazil config), etc that may not be CodeWhisperer inline completion languages

Comment (zixlin7) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/ProjectContextProvider.kt:280]: ok then we don't need to check for language, I can take it out

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/ProjectContextProvider.kt (hunk 3 above)
Comment (zixlin7) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/ProjectContextProvider.kt:293]: in telemetry we are using mb

**Review by leigaol — APPROVED** [review]
Please confirm that the CI failure is not relevant to this change

**Comment by rli** [comment]
241 failure is container hitting OOM

**Pull request merged**